### PR TITLE
Date Manipulation and Duration

### PR DESCRIPTION
**Pull request by devpaul**
A Date Object based on https://github.com/dojo/dojo2/blob/master/src/DateObject.ts and a Duration used for describing a period of time.

### CODE REVIEW

File: src/Duration.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	constructor(... rest: number[]) {
+		var time = 0;
+		for (var i = 0; i < rest.length; i++) {
+			time += rest[i] * unitValues[i];
+		}
+		this.time = time;
+	}
+
+	get milliseconds(): number {
+		return this.time % Duration.SECOND;
```
Comment (bryanforbes) [inline src/Duration.ts:24]: Why is this doing a mod?

File: src/Duration.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+	get hours(): number {
+		return convertUnits(this.time % Duration.DAY, Duration.HOUR);
+	}
+
+	get days(): number {
+		return convertUnits(this.time, Duration.DAY);
+	}
+
+	add(value: Duration): Duration {
```
Comment (devpaul) [inline src/Duration.ts:43]: include support for add(value: number): Duration

File: src/Duration.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+		return new Duration(this.time + value.time);
+	}
+
+	difference(value: Duration): Duration {
```
Comment (devpaul) [inline src/Duration.ts:47]: include support for difference(value: number): Duration

File: src/Duration.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+		return new Duration(value.time - this.time);
+	}
+
+	compare(value: Duration): number {
```
Comment (devpaul) [inline src/Duration.ts:51]: include support for compare(value: number): number

File: src/Duration.ts (hunk 1 above)
Comment (bryanforbes) [inline src/Duration.ts:24]: I talked with @devpaul and the mod is to make the `duration.milliseconds` API similar to `dateObject.milliseconds` which makes perfect sense.

File: src/Duration.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
+var unitValues = [1, 1000, 60000, 3600000, 86400000];
+
+function convertUnits(time: number, unitMilliseconds: number) {
+	return (time > 0 ? 1 : -1) * Math.floor(Math.abs(time) / unitMilliseconds);
+}
+
+export default class Duration {
+	public static SECOND = unitValues[1];
```
Comment (devpaul) [inline src/Duration.ts:8]: move public statics into export const enum Interval { Millisecond, Second = 1000, ... }

File: src/DateObject.ts [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+
+	get dayOfWeek(): number {
+		return this._date.getDay();
+	}
+	get timezoneOffset(): number {
+		return this._date.getTimezoneOffset();
+	}
+
+	add(value: number): DateObject;
+	add(value: Duration): DateObject;
```
Comment (devpaul) [inline src/DateObject.ts:260]: We should add DateObject.subtract()

File: src/DateObject.ts [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+		}
+		else {
+			// Attempt to put value into a DateObject and translate to a time
+			result.time += new DateObject((<any> value)).time;
+		}
+
+		return result;
+	}
+
+	difference(value: number): Duration;
```
Comment (devpaul) [inline src/DateObject.ts:290]: Consider renaming difference() to until() to match with Duration.until() and disambiguate from subtract(). difference() is ambiguous in general practice when returning a Duration once the difference goes beyond the border of a month or leap year. until() is more direct in answering date1.until(date2) -> duration.

**Pull request closed without merging**